property x0shift, x1shift, y0shift, y1shift for adjusting the shape coordinates

my-tien opened the pull request:
This property adjusts the shape coordinates if xref/yref references a (multi-)category axis.

A shape at position x0/x1 = "A"
![image](https://github.com/plotly/plotly.js/assets/3898364/5753a78f-dc3c-4d56-a434-425a4fc364c9)

with x0shift = 0.5
![image](https://github.com/plotly/plotly.js/assets/3898364/bed614e2-a96d-46f5-ab1b-ca54781e291f)

<details>
<summary>Disclaimer</summary>
I am required to add that…

the software is provided "as is", without warranty of any kind, express or implied, including but not limited to the warranties of merchantability, fitness for a particular purpose and noninfringement. in no event shall the authors or copyright holders be liable for any claim, damages or other liability, whether in an action of contract, tort or otherwise, arising from, out of or in connection with the software or the use or other dealings in the software.
</details >

archmoj commented on src/components/selections/attributes.js:
```diff
@@ -53,7 +53,28 @@ module.exports = overrideAll(templatedArray('selection', {
         description: 'Sets the selection\'s end y position.'
     },
-
+    x_shift: {
```
Let's drop the underscore from the name of the attribute.
We do have attributes e.g. `plot_bgcolor` and `paper_bgcolor`; but we don't want to add more.

archmoj commented on src/components/selections/attributes.js:
```diff
@@ -53,7 +53,28 @@ module.exports = overrideAll(templatedArray('selection', {
+        editType: 'calc',
+        description: [
+            'Only relevant if xref is a (multi-)category axes. Shifts x0 and x1 by a fraction of',
+            'the reference unit.'
```
What happens in the case of `axis.type: 'date'`?

alexcjohnson replied:
> Shifts x0 and x1

What if you wanted to shift x0 and x1 differently? Like to surround a particular bar with a rect?

archmoj commented on src/components/selections/defaults.js:
```diff
@@ -49,6 +49,8 @@ function handleSelectionDefaults(selectionIn, selectionOut, fullLayout) {
     coerce('line.color');
     coerce('line.width');
     coerce('line.dash');
+    coerce('xshift');
```
Please check the axis `type` before coercing.

my-tien replied:
done 👍

archmoj commented on src/components/shapes/helpers.js:
```diff
@@ -208,23 +224,22 @@ exports.getPathString = function(gd, options) {
-        x0 = x2p(options.x0);
-        x1 = x2p(options.x1);
+        x0 = x2p(options.x0) + shiftUnitX * options.xshift;
+        x1 = x2p(options.x1) + shiftUnitX * options.xshift;
```
Many duplicates of `shiftUnitX * options.xshift` here.
Let's put it in a variable and use it in those places.

my-tien replied:
done 👍

archmoj commented on src/components/selections/attributes.js:
```diff
@@ -54,6 +54,29 @@ module.exports = overrideAll(templatedArray('selection', {
+            'the reference unit.'
+        ].join(' ')
+    },
+
```
We should also add these to `newselection` and test it.

archmoj commented on test/image/mocks/zzz_shape_shift_horizontal.json:
```diff
@@ -0,0 +1,87 @@
+{
+    "data": [
+        {
```
Let's add a `scatter` and a `scattergl` trace with several points to test if they are selected properly.

my-tien replied:
I added a scatter trace to shape_shift_vertical where there is also a selection.
When I tried to add scattergl, it wasn't selected. But this was also the case when I tried it out on master (same mock, with type "scattergl" instead of "scatter"). Am I missing something?

archmoj commented on test/image/mocks/zzz_shape_shift_vertical.json:
```diff
@@ -0,0 +1,65 @@
+                "y1": 0.5,
+                "xshift": -0.25,
+                "yref": "paper"
+            }
```
Is this also working for a path?

my-tien replied:
No, this should only do something when using x0, x1, y0, y1

archmoj commented:
Interesting PR for shapes and selections.
But I started wondering perhaps we should add a `categoryshift` option to Cartesian axes instead?

archmoj commented on src/components/shapes/attributes.js:
```diff
@@ -220,7 +220,28 @@ module.exports = templatedArray('shape', {
+            'Only relevant if yref is a (multi-)category axes. Shifts y0 and y1 by a fraction of',
+            'the reference unit.'
+        ].join(' ')
+    },
```
Please have a look at `text_on_shapes_basic` mock where you could draw new shapes as described by the `json` spec on the graph.
As per my comment https://github.com/plotly/plotly.js/pull/7005#issuecomment-2142381614
I am still not convinced that these new attributes belong to shape and selection not the axis.

alexcjohnson replied:
Yeah seems weird for newshape. It would need to be factored into the calculation for converting what you draw to coordinates, so what’s the purpose? I’d suggest omitting it until we see a clear use case.

archmoj replied:
I am thinking here that instead of just shifting values, what users want is to round them to categories (i.e. integers) and then use these offsets to declare the start and end. A feature like `snap to grid`. 
This would make selection on categories easier and could have application in other axis types too.

my-tien replied:
Hi @archmoj is it possible to postpone this to a later PR? I didn't anticipate this feature and haven't looked at newshape/newselection before, so I'm not quite sure how to approach this feature atm.

archmoj commented:
I am wondering if adding a `categoryselectionstart`, `categoryselectionend`, `categoryshapestart` and  `categoryshapeend` to cartesian axes is a better option?

archmoj commented:
Let's ask @alexcjohnson to review :mag_right:

alexcjohnson commented on test/image/mocks/zzz_shape_shift_horizontal.json:
```diff
@@ -0,0 +1,87 @@
+                  "color": "black",
+                  "width": 3.0
+                },
+                "y0": ["A", "C"],
```
Yeah to know that this surrounds A,D you have to know that C is right before D, which is fragile. Much better in this case is if we had a way to specify two shifts per dimension.

my-tien commented:
> I am wondering if adding a `categoryselectionstart`, `categoryselectionend`, `categoryshapestart` and `categoryshapeend` to cartesian axes is a better option?

Here is a PR with this suggestion for shapes only: #7010

archmoj commented:
Considering https://github.com/plotly/plotly.js/pull/7010#issuecomment-2147726890 discussion,
two attributes are needed for start and the end i.e. instead of `xshift` which is already implemented we need `x0shift` and `x1shift`.

archmoj commented on src/components/selections/attributes.js:
```diff
@@ -54,6 +54,29 @@ module.exports = overrideAll(templatedArray('selection', {
+        valType: 'number',
+        dflt: 0,
+        min: -0.5,
+        max: 0.5,
```
We could/should allow ranges beyond [-0.5, 0.5] so that one could highlight a histogram or a bar with `width: 1` using a value like 0.7.

my-tien replied:
done 👍

archmoj commented on src/components/shapes/attributes.js:
```diff
@@ -169,7 +169,17 @@ module.exports = templatedArray('shape', {
+        valType: 'number',
+        dflt: 0,
+        min: -0.5,
+        max: 0.5,
```
We could/should allow ranges beyond [-0.5, 0.5] so that one could highlight a histogram or a bar with width: 1 using a value like 0.7.

my-tien commented on src/components/shapes/defaults.js:
```diff
@@ -75,6 +75,9 @@ function handleShapeDefaults(shapeIn, shapeOut, fullLayout) {
         var pos2r;
         var r2pos;
 
+        coerce('xshift');
```
done 👍

archmoj commented on draftlogs/7005_add.md:
```diff
@@ -0,0 +1 @@
+ - Add property x_shift/y_shift for shapes/selections referencing (multi-)category axes [[#7005](https://github.com/plotly/plotly.js/pull/7005)]
```
Let's update this description using the final attribute names.

my-tien replied:
done

archmoj commented on src/components/selections/attributes.js:
```diff
@@ -54,6 +54,51 @@ module.exports = overrideAll(templatedArray('selection', {
+        valType: 'number',
+        dflt: 0,
+        min: -1,
+        max: 1,
```
Let's mention in the descriptions that *-0.5* is the start of the category and *0.5* is the end of it.

my-tien replied:
done

archmoj commented on src/components/shapes/helpers.js:
```diff
@@ -208,23 +231,22 @@ exports.getPathString = function(gd, options) {
         var xAnchorPos = x2p(options.xanchor);
-        x0 = xAnchorPos + options.x0;
-        x1 = xAnchorPos + options.x1;
+        x0 = xAnchorPos + options.x0 + xShiftStart;
```
It would be nice if you could declare a centralize function to handle these adjustments here and in other files if used.

Considering my comment about possibility of snapping to the (category) grid [here](https://github.com/plotly/plotly.js/pull/7005/files#r1633371678), you may need to round the current value before adding start and end shifts. Having a centralized function would help testing that idea later.

Thank you!

my-tien replied:
I moved the logic for calculating the shift in pixels into a centralized function, I did not yet include any rounding.

my-tien commented on src/components/selections/attributes.js:
```diff
@@ -54,6 +54,51 @@ module.exports = overrideAll(templatedArray('selection', {
+        max: 1,
+        editType: 'calc',
+        description: [
+            'Only relevant if xref is a (multi-)category axes. Shifts x0 by a fraction of the',
```
done

archmoj commented:
#6347 seems related to this PR. No?

archmoj commented on test/jasmine/tests/shapes_test.js:
```diff
@@ -1504,18 +1510,18 @@ describe('Test shapes', function() {
-        var x2p = helpers.getDataToPixel(gd, xa);
-        var y2p = helpers.getDataToPixel(gd, ya, true);
+        var x2p = helpers.getDataToPixel(gd, xa, layoutShape.xshift);
+        var y2p = helpers.getDataToPixel(gd, ya, layoutShape.yshift, true);
```
Do we have `xshift` and `yshift` on the `layoutShape` here?

my-tien replied:
actually no, the function that creates the shapes does not set a shift. Should I just add xshift and yshift 0 to the ` setupLayout` function?

archmoj replied:
If required yes.

my-tien replied:
Yeah, so right now they just evaluate to "undefined" because they are not set, which the function handles fine. But yeah, might be clearer.

my-tien replied:
Nvm, now I get your question and updated the code. I just overlooked the remaining places where xshift and yshift were still used instead of x0/x1/y0/y1shift.

stephprobst commented:
> #6347 seems related to this PR. No?

Yes. The current PR may be part of the groundwork required for #6347. Do you think any additional changes are required to merge the current PR though?

archmoj commented:
> > #6347 seems related to this PR. No?
> 
> Yes. The current PR may be part of the groundwork required for #6347. Do you think any additional changes are required to merge the current PR though?

@stephprobst In your use cases do you want to move the coordinates by either -0.5 or +0.5 only? 
Or you want to have control for any ratio e.g. between -1 and +1?
Also we need to know if you are interested to snap to those values or just moving without any rounding?

archmoj commented:
Thanks for the feedback @stephprobst.
It's good to know that -0.5 and 0.5 values are sufficient in your use cases.
With that I think rounding start and end values gives us a better API.
Also in case of selection (and new selection) I think rounding would be useful both for rectangular selections as well as `lasso` selection.
What do you think?

my-tien commented:
Do you mean that plotly should internally round the shift values provided by the user to the closest 0.5 increment/decrement? Wouldn't it then be a little confusing to be able to enter any float value between -1 and 1 instead of only allowing -1, -0.5, 0, 0.5 and 1?

stephprobst commented:
**On the discussion of floating point precision vs. rounding:** We currently only have a use case for +/- 0.5, but I'm not sure other floating point values won't be useful in the future. At the same time I see the value of rounding. Would it be possible to say we do **both**? Finish the current PR with the floating point precision and add additional keywords `roundup` and `rounddown` at a later stage?

**On the discussion of "add property to axis" vs. "add property to shape":** We would strongly prefer this property to be available on the individual shape, as it may very well be, that the user wants to combine a vertical line that separates categories with other shapes, that should be linked to the center of the category. Thanks!

LiamConnors commented on src/components/shapes/attributes.js:
```diff
@@ -169,7 +169,28 @@ module.exports = templatedArray('shape', {
+        min: -1,
+        max: 1,
+        editType: 'calc',
+        description: [
+            'Only relevant if `xref` is a (multi-)category axes. Shifts `x0` by a fraction of the',
+            'reference unit. E.g. -0.5 corresponds to the start of the category and 0.5 corresponds to the end of the category.'
+        ].join(' ')
```
Does `(multi-) category` refer to one of the types as defined here: https://plotly.com/python/reference/layout/xaxis/#layout-xaxis-type

Is the `(multi-)`  to convey that it applies only to axes with the types "multicategory" or "category"

my-tien replied:
Yes exactly, it applies to axes with `type` "multicategory" or `type` "category".

archmoj commented:
@my-tien There are not conflicts with master; but please fetch & merge `upstream/master` into this branch.

archmoj commented on src/components/selections/attributes.js:
```diff
@@ -54,6 +54,51 @@ module.exports = overrideAll(templatedArray('selection', {
+    x0shift: {
+        valType: 'number',
+        dflt: 0,
+        min: -1,
```
For `selections` limiting the range to be between -0.5 and 0.5 is safer IMHO.

my-tien replied:
As discussed, I removed shift attributes from selections for now.

archmoj commented:
@stephprobst Do you have a use case for adding these options to selections?
If no, I'd suggest not to implement it in this PR.

stephprobst commented:
> @stephprobst Do you have a use case for adding these options to selections? If no, I'd suggest not to implement it in this PR.

@archmoj: No. We're not using selections at the moment and probably won't in the foreseeable future. I'm ok with skipping the implementation in this PR.

archmoj commented:
> > @stephprobst Do you have a use case for adding these options to selections? If no, I'd suggest not to implement it in this PR.
> 
> @archmoj: No. We're not using selections at the moment and probably won't in the foreseeable future. I'm ok with skipping the implementation in this PR.

@my-tien Let's simplify this PR by adding shift options to `shapes` only i.e. no `newshape` and no `selections` with no rounding.
Thank you!

archmoj commented:
Thanks for the simplification.
Please update the PR title and description.
Also `zzz_shape_shift_vertical` image test is failing now.

archmoj commented:
@stephprobst Do you have a use case for `date` axes?
If so it could be possible to add these options for other axis types.

archmoj commented on src/components/shapes/helpers.js:
```diff
@@ -279,3 +283,19 @@ function convertPath(options, x2p, y2p) {
+    if(axis && shift) {
+        var isVertical = axis._id.charAt(0) === 'y';
+        if(axis.type === 'category' || axis.type === 'multicategory') {
+            if(isVertical) {
```
Please test these on reversed x and y axes too.
Thank you!

my-tien replied:
thx, it didn't actually work and the fix resulted in simpler code.

stephprobst commented:
> @stephprobst Do you have a use case for `date` axes? If so it could be possible to add these options for other axis types.

@archmoj No, I don't think this is relevant for other axes. The rationale for this property is to allow exact positioning of a shape based on the axis value. For categorical axes this is needed, because it allows the combination of categorical and numerical values (e.g. country "Canada" plus a shift of 0.5 to position the shape at the end of the categories interval on the axis). On numerical or date axes this isn't really needed, since the user can directly select the precise numerical or date value (e.g. "2023-12-31" to position a shape at the end of year 2023).

My proposal would be to skip the implementation for date and numeric axes.

archmoj commented on src/components/shapes/calc_autorange.js:
```diff
@@ -77,15 +73,35 @@ function calcPaddingOptions(lineWidth, sizeMode, v0, v1, path, isYAxis) {
+    var shiftStart = 0;
+    var shiftEnd = 0;
+
+    var convertVal = isCategory ? ax.r2c : ax.d2c;
```
This could be simplified now that we only support category axis types.

my-tien replied:
Do you mean, moving the shiftStart/shiftEnd into the convertVal function?
I didn't do that because in the case of sizemode `pixel` I think I need to apply convertVal but not the shift?
Not sure though, is this correct?

archmoj replied:
Is there a case where the value of `isCategory` is *false*?

my-tien replied:
yes, this function is called for shapes on all axes, so it requires a category check

archmoj replied:
Thanks for the note.
Resolving.

archmoj commented on src/components/shapes/helpers.js:
```diff
@@ -53,7 +53,7 @@ exports.extractPathCoords = function(path, paramsToUse, isRaw) {
 };
 
-exports.getDataToPixel = function(gd, axis, isVertical, refType) {
+exports.getDataToPixel = function(gd, axis, shift, isVertical, refType) {
```
We could drop the 4th argument (isVertical) and get it from the `axis._id.chartAt(0) === 'y'`.

my-tien replied:
This is unfortunately not possible, because it could also be vertical if axis is undefined (when the xref or yref is paper).
That is probably why this param was there before already.

archmoj replied:
Thanks for the note.
Resolving...

archmoj commented:
Please also test `texttemplate` on your mocks to show the slopes similar to those in the `text_on_shapes_texttemplate` mock. 
Thank you!

my-tien commented:
> Please also test `texttemplate` on your mocks to show the slopes similar to those in the `text_on_shapes_texttemplate` mock. Thank you!

I noticed that I didn't account for the shift for the texttemplate. Fixed that and added a test shape with slope and xcenter.

my-tien commented:
> When `shape.editable` is set to true, clicking the shape and editing the position of a vertex result in the wrong positioning of the updated shape.

Fixed

archmoj commented on src/components/shapes/attributes.js:
```diff
@@ -169,7 +169,28 @@ module.exports = templatedArray('shape', {
+        max: 1,
+        editType: 'calc',
+        description: [
+            'Only relevant if `xref` is a (multi-)category axes. Shifts `x0` by a fraction of the',
```
We should firstly describe what the attribute does.
@LiamConnors What's your suggestion for the attribute description?

LiamConnors commented on src/components/shapes/attributes.js:
```diff
@@ -169,7 +169,28 @@ module.exports = templatedArray('shape', {
+        max: 1,
+        editType: 'calc',
+        description: [
+            'Only relevant if `xref` is a (multi-)category axes. Shifts `x0` by a fraction of the',
+            'reference unit. E.g. -0.5 corresponds to the start of the category and 0.5 corresponds to the end of the category.'
```
Shifts `x0` away from the centre of the category when `xref` is a "category" or "multicategory" axis. 
-0.5 corresponds to the start of the category and 0.5 corresponds to the end of the category.

What do you think of this @archmoj. Is it technically correct?

archmoj replied:
@LiamConnors this is correct.
@my-tien Please adjust the descriptions and update the schema and I think we should be good to merge this PR.
Thank you!

my-tien replied:
Thanks for turning it into something comprehensible :)
I just changed this part to use asterisk like you do in other descriptions:
> \*category\* or \*multicategory\* axis

archmoj reviewed (approved):
Thanks @my-tien  for the PR :1st_place_medal: 
:dancer:
